src/metabase/models/params/custom_values.clj

```diff
@@ -0,0 +1,166 @@
+      (first (filter (fn [field-info]
+                       (and (= (:name field-info)
+                               id-or-name)
+                            (= (:base-type options)
+                               (:base_type field-info))))
+                     result-metadata)))))
+
+(defn- values-from-card-query
+  [card value-field query]
+  (let [value-base-type (:base_type (field->field-info value-field (:result_metadata card)))]
+    {:database (:database_id card)
+     :type     :query
+     :query    (merge
+                 {:source-table (format "card__%d" (:id card))
+                  :breakout     [value-field]
+                  :limit        *max-rows*}
+                 {:filter [:and
+                           [(if-not (isa? value-base-type :type/Text)
+                              :not-null
+                              :not-empty)
+                            value-field]
+                           (when query
+                             (if-not (isa? value-base-type :type/Text)
+                               [:= value-field query]
+                               [:contains [:lower value-field] (u/lower-case-en query)]))]})
+     :middleware {:disable-remaps? true}}))
+
+(mu/defn values-from-card
+  "Get distinct values of a field from a card.
+
+  (values-from-card 1 [:field \"name\" nil] \"red\")
+  ;; will execute a mbql that looks like
+  ;; {:source-table (format \"card__%d\" card-id)
+  ;;  :fields       [value-field]
+  ;;  :breakout     [value-field]
+  ;;  :filter       [:contains [:lower value-field] \"red\"]
+  ;;  :limit        *max-rows*}
+  =>
+  {:values          [\"Red Medicine\"]
+  :has_more_values false}
+  "
+  ([card value-field]
+   (values-from-card card value-field nil))
+
+  ([card            :- (ms/InstanceOf Card)
+    value-field     :- ms/Field
+    query           :- [:any]]
+   (let [mbql-query   (values-from-card-query card value-field query)
+         result       (qp/process-query mbql-query)
+         values       (map first (get-in result [:data :rows]))]
+     {:values          values
+      ;; if the row_count returned = the limit we specified, then it's probably has more than that
+      :has_more_values (= (:row_count result)
+                          (get-in mbql-query [:query :limit]))})))
+
+(defn card-values
+  "Given a param and query returns the values."
+  [{config :values_source_config :as _param} query]
+  (let [card-id (:card_id config)
+        card    (db/select-one Card :id card-id)]
+    (values-from-card card (:value_field config) query)))
+
+(defn- check-can-get-card-values?
+  [card value-field]
+  (cond
+    (:archived card)
+    false
+
+    (not (field->field-info value-field (:result_metadata card)))
+    false
+
+    :else true))
+
+;;; --------------------------------------------- Putting it together ----------------------------------------------
+
+(defn parameter->values
```

Comment:
this and `check-can-get-card-values?` are the only 2 new functions. the rest are just copied over.